File: android/core/src/main/java/com/americangroupllc/drift/core/networking/SupabaseClient.kt
package com.americangroupllc.drift.core.networking

import io.ktor.client.HttpClient
import io.ktor.client.engine.HttpClientEngine
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.headers
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.HttpStatusCode
import io.ktor.http.URLBuilder
import io.ktor.http.appendPathSegments
import io.ktor.http.contentType
import io.ktor.http.ContentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

/**
 * Thin Ktor wrapper around the bits of the Supabase REST API Drift uses.
 * Mirrors the surface of `SupabaseClient.swift`.
 */
class SupabaseClient(
    val baseUrl: String,
    val anonKey: String,
    engine: HttpClientEngine = CIO.create(),
) {
    private var jwt: String? = null

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = false }

    private val client = HttpClient(engine) {
        install(ContentNegotiation) {
            json(json)
        }
        expectSuccess = false
    }

    fun setJwt(token: String?) { jwt = token }

    suspend fun rawGet(path: String, query: Map<String, String> = emptyMap()): HttpResponse =
        client.get(
            URLBuilder(baseUrl).apply {
                appendPathSegments(path.trim('/').split('/'))
                query.forEach { parameter(it.key, it.value) }
            }.build()
        ) { headers { applyHeaders(this) } }

    suspend inline fun <reified Body : Any> postJson(
        path: String, body: Body,
    ): HttpResponse = client.post(buildUrl(path)) {
        headers { applyHeaders(this) }
        contentType(ContentType.Application.Json)
        setBody(body)
    }

    suspend inline fun <reified Body : Any> invokeFunction(
        name: String, body: Body,
    ): HttpResponse = postJson("functions/v1/$name", body)

    @PublishedApi
    internal fun buildUrl(path: String): String =
        baseUrl.trimEnd('/') + "/" + path.trimStart('/')

    @PublishedApi
    internal fun applyHeaders(b: io.ktor.http.HeadersBuilder) {
        b.append("apikey", anonKey)
        b.append("Authorization", "Bearer ${jwt ?: anonKey}")
        b.append("Accept", "application/json")
    }

    fun close() { client.close() }
}

class SupabaseHttpException(val status: HttpStatusCode, val body: String) :
    RuntimeException("HTTP ${status.value}: $body")

File: apps/DriftDate/shared/DriftCore/Sources/DriftCore/Networking/SupabaseClient.swift
import Foundation

/// A thin URLSession-based wrapper around the bits of the Supabase REST
/// API that Drift uses. We deliberately avoid `supabase-swift` to keep
/// the surface tiny and easy to mock in tests.
///
/// Reads `SUPABASE_URL` and `SUPABASE_ANON_KEY` from `Bundle.main.infoDictionary`
/// (set via `Info.plist` build settings) so the same binary points at
/// dev / staging / prod via a build-time substitution.
public final class SupabaseClient {
    public let baseURL: URL
    public let anonKey: String
    private let session: URLSession
    private var jwt: String?

    public init(baseURL: URL, anonKey: String, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.anonKey = anonKey
        self.session = session
    }

    public static func fromBundle(bundle: Bundle = .main, session: URLSession = .shared) -> SupabaseClient? {
        let url = (bundle.infoDictionary?["SUPABASE_URL"] as? String).flatMap(URL.init(string:))
        let key =  bundle.infoDictionary?["SUPABASE_ANON_KEY"] as? String
        guard let url, let key, !key.isEmpty else { return nil }
        return SupabaseClient(baseURL: url, anonKey: key, session: session)
    }

    public func setJWT(_ token: String?) {
        self.jwt = token
    }

    // MARK: - REST

    public func get<T: Decodable>(
        _ path: String,
        query: [String: String] = [:],
        as type: T.Type = T.self
    ) async throws -> T {
        var comps = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if !query.isEmpty {
            comps.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }
        var req = URLRequest(url: comps.url!)
        applyHeaders(&req)
        return try await execute(req)
    }

    public func post<Body: Encodable, T: Decodable>(
        _ path: String,
        body: Body,
        as type: T.Type = T.self
    ) async throws -> T {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = "POST"
        applyHeaders(&req)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder.driftDefault.encode(body)
        return try await execute(req)
    }

    public func invokeFunction<Body: Encodable, T: Decodable>(
        _ name: String,
        body: Body,
        as type: T.Type = T.self
    ) async throws -> T {
        try await post("functions/v1/\(name)", body: body, as: T.self)
    }

    // MARK: - Internals

    private func applyHeaders(_ req: inout URLRequest) {
        req.setValue(anonKey, forHTTPHeaderField: "apikey")
        if let jwt {
            req.setValue("Bearer \(jwt)", forHTTPHeaderField: "Authorization")
        } else {
            req.setValue("Bearer \(anonKey)", forHTTPHeaderField: "Authorization")
        }
        req.setValue("application/json", forHTTPHeaderField: "Accept")
    }

    private func execute<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        guard (200..<300).contains(http.statusCode) else {
            throw SupabaseError.http(status: http.statusCode, body: data)
        }
        if T.self == EmptyResponse.self {
            return EmptyResponse() as! T
        }
        return try JSONDecoder.driftDefault.decode(T.self, from: data)
    }
}

public struct EmptyResponse: Decodable {}

public enum SupabaseError: Error, Equatable {
    case http(status: Int, body: Data)
}

extension JSONEncoder {
    public static let driftDefault: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        e.dateEncodingStrategy = .iso8601
        return e
    }()
}

extension JSONDecoder {
    public static let driftDefault: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        d.dateDecodingStrategy = .iso8601
        return d
    }()
}
